lib/controller.dart

```diff
@@ -300,463 +280,162 @@ void loadGeoMapsData() async {
-    _analyseOptions = model.AnalyseOptions(0, true, false, true);
-    _filters = [];
-  }
-
-  var uri = Uri.parse(html.window.location.href);
-  var queryParams = uri.queryParameters;
-  var chartOptions = convert.jsonDecode(queryParams['chartOptions'] ?? '{}');
-  _analyseOptions.updateFrom(chartOptions);
-
-  var queryTabLabel = _config.tabs[_analyseOptions.selectedTabIndex].label;
-
-  var tabLabels =
-      _config.tabs.asMap().map((i, t) => MapEntry(i, t.label)).values.toList();
-  view.renderAnalysisTabs(tabLabels, queryTabLabel);
-  view.renderChartOptions(
+  view.AnalyseTabsViews(_config.tabs.map((t) => t.label).toList(),
+      _analyseOptions.selectedTabIndex);
+  view.ChartOptionsView(
       _analyseOptions.dataComparisonEnabled,
       _analyseOptions.normaliseDataEnabled,
       _analyseOptions.stackTimeseriesEnabled);
+  _dataFilterView = view.DataFiltersView();
+  _dataFilterView.update(_filtersMap[_analyseOptions.selectedTabIndex]);
 
-  _computeFilterDropdownsAndRender(
-      _config.tabs[_analyseOptions.selectedTabIndex].filters);
-  _computeChartDataAndRender();
+  _initialiseCharts();
+  _computeCharts();
+  _updateCharts();
 }
 
-void _computeFilterDropdownsAndRender(List<model.Filter> filters) {
-  filters = filters ?? [];
-
-  _filters = [];
-  filters.forEach((f) {
-    var filterOptions = <String>[];
-    var defaultValue = DEFAULT_FILTER_SELECT_VALUE;
-    if (f.type == model.DataType.string) {
-      filterOptions = _uniqueFieldValues[f.data_path][f.key]
-          .map((e) => e.toString())
-          .toList()
-            ..add(DEFAULT_FILTER_SELECT_VALUE);
-    } else if (f.type == model.DataType.datetime) {
-      filterOptions = [];
-      switch (f.data_path) {
-        case model.DataPath.message_stats:
-          // todo: check across all _messageStats, not just the first
-          var dates = _messageStats.values.first.keys.toList()..sort();
-          filterOptions = [dates.first, dates.last];
-          var splitCharacter = 'T';
-          if (!dates.first.toString().contains(splitCharacter)) {
-            splitCharacter = ' ';
+// Filter options
+void _computeFilterOptions() {
+  for (var i = 0; i < _config.tabs.length; ++i) {
+    var currentTab = _config.tabs[i];
+    for (var j = 0; j < currentTab.filters.length; ++j) {
+      var currentFilter = currentTab.filters[j];
+      if (_dataCollections[currentFilter.data_collection].isEmpty) continue;
+      switch (currentFilter.type) {
+        case model.DataType.datetime:
+          var collectionToIterate =
+              _dataCollections[currentFilter.data_collection].values;
+          var dates = collectionToIterate
+              .map((e) => e[currentFilter.key])
+              .toList()
+                ..sort();
+          _filtersMap[i][j].options = [dates.first, dates.last];
+          if (_filtersMap[i][j].value == '') {
+            _filtersMap[i][j].value =
+                '${dates.first.split("T")[0]}_${dates.last.split("T")[0]}';
```

Comment:
it seems that you need to do this a lot - can we maybe create a `getDate(String datetime)` that implements this logic of extracting the date?

Reply:
Yep sure, will refactor to a separate function